testing/matchers/be_successful_state.go

```diff
@@ -0,0 +1,64 @@
+package matchers
+
+import (
+	"fmt"
+
+	"github.com/onsi/gomega/types"
+	"github.com/pivotal-cf/brokerapi/domain"
+)
+
+const MaxConcurrentFailures = 2
+
+// BeSuccessState is a gomega custom matcher for checking that the result of
+// receiving from a `chan domain.LastOperationState` is a successful state. It
+// mostly works like the `Receive(Equal(state))` matcher, but with the addition
+// that it understands how to bail out early from terminal conditions, making
+// it useful for use in Eventually calls polling the state chan
+func BeSuccessState() types.GomegaMatcher {
+	return &haveLastOperationStateMatcher{
+		expectedState: domain.Succeeded,
+	}
+}
+
+type haveLastOperationStateMatcher struct {
+	expectedState   domain.LastOperationState
+	inTerminalState bool
+	failures        int
+}
+
+func (matcher *haveLastOperationStateMatcher) Match(actual interface{}) (success bool, err error) {
+	ch, ok := actual.(chan domain.LastOperationState)
+	if !ok {
+		return false, fmt.Errorf("HaveLastOperationState matcher expects an chan domain.LastOperationState")
+	}
+	select {
+	case actualState, ok := <-ch:
+		if !ok { // channel got closed, bail out
+			matcher.inTerminalState = true
+			return false, nil
+		} else if actualState == domain.Failed { // if we see multiple failures in a row bail out
```

Comment:
The channel will _always_ initially have a failed state in it.... because we set up the channel before we have provisioned anything.

This could be avoided by only setting up the chan, after creation is complete .... but that was messier and hard to read ... so this lets you avoid that.

Some work could probably be done to make that nicer, but my main aim here was to speed up the feedback loop and makes things easier to read.

Reply:
huh, it feels weird to me that you get `domain.Failed` from the last operation endpoint if you haven't tried yet. 😕  but yeah if that's the behaviour then ugh fine i guess, this is the way it has to be